### ISSUE TITLE
Documentation

### ISSUE THREAD

**Issue by jmurzy** [jmurzy/react-router-native#5]
One of the the primary goals of this project is to keep its API as close to [react-router](https://github.com/reactjs/react-router/tree/master/docs) as makes sense. However, navigation on native platforms is a broader topic and we will probably have to introduce new concepts and options to accommodate for richer animations, the lack of hyperlinks, et al.

So far we've only introduced two additional router configuration components, `<StackRoute>` and `<TabsRoute>`. Also, two new props, `overlayComponent` and `transition`, have been added to ``<Route>` and its siblings.

We should also document how to create custom transition animations using the `transitionRegistry`. ⚠️

:beer:

**Comment by jonathanglasmeyer** [comment]
Quick question. react-router allows specifying the routes configuration as a plain JavaScript object. I would assume that this is not possible for this library. Is that correct?

**Comment by jmurzy** [comment]
@jonathanewerner It is possible to configure routes using a plain object. We have a hard dependency on react-router. I'd say it supports pretty much everything react-router does except for a few props that are [not yet implemented](https://github.com/jmurzy/react-router-native/blob/800622777e0dac89461e378d7e6d4e0d37872215/modules/Route.js#L31-L33).

🍺

**Comment by dchambers** [comment]
Hi @jmurzy,

I'm trying to be self sufficient here, and would love to be able to help write some documentation, but I'm still struggling to understand how `<StackRoute>`, `<TabsRoute>`,  `component`, `overlayComponent`, etc actually work.

To simplify things I decided to modify the example you provide so that it uses a modified form of the markup shown in the 'Usage' section rather than the 'Advanced Usage' section, as follows:

``` jsx
const component = (name) => () => <Text>{name}</Text>;

const routes = (
  <Router history={nativeHistory} addressBar>
    <TabsRoute
      path="app"
      component={component('App')}
      transition="horizontal-pager"
    >
      <Route path="/" component={component('About')} overlayComponent={component('AboutHeader')} />
      <Route path="users" component={component('Users')} overlayComponent={component('UsersHeader')}>
        <Route path="/user/:userId" component={component('User')} />
      </Route>
    </TabsRoute>
    <Route path="*" component={component('NoMatch')} />
  </Router>
);
```

These are the results I get:
- `/` (App)
- `/user` (NoMatch)
- `/user/fred` (App)
- `/foo` (NoMatch)

At no point do I see 'About', 'AboutHeader', 'Users', 'UserHeader', or 'User'. What am I doing wrong or misunderstanding here?

**Comment by peterp** [comment]
One of the things that I've come to understand is just because the routes are nested doesn't mean that they're relative, e.g.: you can still define an absolute path in a nested route, as such your routes can be accessed as follows:
- `app` (App)
- `/` (App + About)
- `app/users` (App + Users)
- `/users/:userId` (App + User)

The easiest way to deal with this is to drop the slashes from the path definitions.

**Comment by dchambers** [comment]
@peterp, thanks for the clarifying the situation with absolute Vs relative routes. That's really helpful 👍 

However, even with that extra information, I'm still only able to see 'App' and 'NoMatch' when trying different routes, and never see 'About', 'Users' or 'User', as you seem to indicate. Is there something else I'm still not getting?

**Comment by peterp** [comment]
@dchambers It's possible that there's something that I'm not understanding as well. Sorry for adding noise to this.

**Comment by jmurzy** [comment]
I apologize for the delay in responding to this.

To understand how route matching works, I strongly recommend the React Router [docs](https://github.com/reactjs/react-router/blob/master/docs/guides/RouteMatching.md). That's what we use under the hood.

In your snippet ^, you've configured a `<TabsRoute>` with two "tabs". The first tab, `About`, can be accessed by navigating to `/`. Second tab, Users, can only be accessed by navigating to `/app/users` since it's defined as a relative path (no leading slash). If you wanted to use an absolute path as in `/users`, you'd have to change your path to be exactly that, `/users`. 

In your current config, going to `/app/users`, however, will only render `Users` and not `User`. If you want to render `User` then you'd have to go to `/user/:userId` which is defined as an absolute path. You could alternatively make that relative by removing the leading slash from the path. In that case, you'd have to navigate to `/app/users/user/0`. I'd also read the React Router docs on `<IndexRoute>`.

Noticed the pattern? So these should render as follows:

> - `/` (App)

**App(About)**

> - `/user` (NoMatch)

**NoMatch**

> - `/user/fred` (App)

**App(Users(User))**

> - `/foo` (NoMatch)

**NoMatch**

More on absolute vs. relative [here](https://github.com/reactjs/react-router/blob/master/docs/guides/RouteConfiguration.md#decoupling-the-ui-from-the-url).

Also keep in mind that, since you nested `User` in `Users`, unless your `Users` component renders the `User` component passed in the `children` prop, the inner component will not be rendered regardless of how a match is found. You have full control over how and where your nested components are rendered.

``` js
const Users = (props) => props.children
```

So given above config, if you navigate to `/user/0` (you cannot omit `0` here unless you make `:userID` [optional](https://github.com/reactjs/react-router/blob/master/docs/guides/RouteMatching.md#path-syntax)), when inspected, your view hierarchy will look like this: App(Users(User)). That is as long as you make sure your parent components include  `props.children` when they `render`.  

All of the above is really the same as React Router.

As for `overlyComponent`, it's used to overlay a (header, footer, or what have you) component over your route `component` and usually contains a fixed navigation bar. That is when a new route under a `<StackRoute>` or `<TabsRoute>` is visited, we use transition animations (which are customizable) to slide in the new component, and you usually don't want your navigation bar to be affected by the transition animations. More context on that in #26.

Let me know if you have any further questions. I can also attempt to walk your through what you're trying to achieve via Discord.

🍺

**Comment by dchambers** [comment]
Hi @jmurzy, thanks so much for taking the time to respond. Just in case it's not clear, the example snippet I provide is your own [simple example](https://github.com/jmurzy/react-router-native#usage), but slightly modified so that I could see which components are actually being rendered.

@peterp already helped me understand that you could use absolute paths in React Router, which I hadn't previously realized. Interestingly, [I tried to get the React Router docs improved](https://github.com/reactjs/react-router/issues/3614) in this area, but their response was "That pattern is no longer encouraged" &mdash; I assume since it's incompatible with Dynamic Routing. So, in light of this, it may be worth updating the example (which I'm happy to do?) not to use absolute paths, since these are no longer documented.

> Also keep in mind that, since you nested User in Users, unless your Users component renders the User component passed in the children prop, the inner component will not be rendered regardless of how a match is found.

Doh, what a stupid oversight on my part!! 😁 🔫  Okay, so when I update the definition of `component` to this:

``` jsx
const component = (name) => {
  const c = ({ children }) => (<View>
    <Text>{name}</Text>
    {children}
  </View>);

  c.propTypes = {
    children: PropTypes.node,
  };

  return c;
};
```

I do start to see somewhat more sensible output (still only embedded header components, and not the actual components themselves), as follows:
- `/` (App, AboutHeader)
- `/user` (NoMatch)
- `/user/fred` (App, UsersHeader)
- `/foo` (NoMatch)

So, I guess the only mystery now is, why don't I see any of the non-overlay components (e.g. `Users` or `User`)?

**Comment by jmurzy** [comment]
Ugh. Sorry about that. That example was adopted from the older React Router README. I was trying to mirror as much as possible to make it all easier for the React Router community. 

But we need a simpler, more complete example in the README. Please feel free to submit a pull request. I greatly appreciate contributions, especially to the documentation and examples, #6.

Aside from the Dynamic Routing issue, I think, discussion here can serve as proof that relative paths are a bit harder to grasp. Declaring your paths explicitly can surely minimize the cognitive effort required to understand the route configuration. So I can see why React Router discourages it, and I agree we should too.

> So, I guess the only mystery now is, why don't I see any of the non-overlay components (e.g. Users or User)?

You also need to set your [styles](https://github.com/jmurzy/react-router-native/blob/master/examples/Aviato/app/components/styles.js#L27-L29) correctly for NavigationExperimental to work its magic. 

``` js
<View style={{flex: 1}}>
   {children}
</View>
```

🍺

**Comment by dchambers** [comment]
Thanks @jmurzy! I'm just about to go to bed now, but will give this a try tomorrow. Assuming all goes well, I'd like to create a documentaton improvement pull-request that explains things in a way that would have been helpful to me, in case that information may also be helpful for others.

Thanks again :+1:

**Comment by dchambers** [comment]
Hi @jmurzy, I've just updated my example as per your recommendation and things are now looking good!

I do have a couple more questions though:
1. Is there actually way to have a header overlay rather than a footer overlay given that both [TabsRouteView](https://github.com/jmurzy/react-router-native/blob/7642d6a01415d2c7f6de7a89a071ac9a130805fd/modules/TabsRouteView.js#L165) and [StackRouteView](https://github.com/jmurzy/react-router-native/blob/7642d6a01415d2c7f6de7a89a071ac9a130805fd/modules/StackRouteView.js#L165) always place the overlay after the scene, and given that React Native doesn't support `flexDirection: 'column-reverse'`?
2. What's the intended difference between `TabsRouteView` & `StackRouteView` &mdash; I've diffed these two files and I only see superficial differences between the two?

I think with these questions answered this should be enough for me to hopefully write something sensible. Thanks again!
